Remove Deprecated Extensions

vince-fugnitto opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md
-->

#### What it does
<!-- Include relevant issues and describe how they are addressed. -->

The pull-request removes deprecated extensions from the repository since these extensions will no longer be maintained and are either replaced by `builtin` or external VS Code extensions:

- [x] `@theia/debug-nodejs`
- [x] `@theia/editorconfig`
- [x] `@theia/java`
- [x] `@theia/java-debug`
- [x] `@theia/merge-conflicts`
- [x] `@theia/python`
- [x] `@theia/textmate-grammars`
- [x] `@theia/tslint`
- [x] `@theia/typescript`

**Prerequisite**

- [x] update [theia-apps](https://github.com/theia-ide/theia-apps) images and electron apps

#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->

Verify that CI successfully passes (build & tests)

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

Signed-off-by: Vincent Fugnitto <vincent.fugnitto@ericsson.com>

akosyakov commented:
We should also review theia docs and theia extension generator to make sure that we don't mention them anymore or replace with corresponding VS Code extension. For instance https://theia-ide.org/docs/composing_applications/ is missing a section how to include support of VS Code extensions.

vince-fugnitto commented:
Placing it as "Ready for Review" so I can anticipate any breakage/merge conflicts :)

akosyakov commented:
Depends on https://github.com/eclipse-theia/theia/issues/6946 and https://github.com/eclipse-theia/theia/issues/6947

vince-fugnitto commented:
@eclipse-theia/core 

Do you think we should merge for the upcoming release (Thursday Feb 27th), or should we hold off till right after the release and potentially address any loose ends we might have missed. My understanding is that we should at least aim for before v1.0.

kittaakos commented on CHANGELOG.md:
```diff
@@ -10,7 +10,7 @@
   - `prepare:references`: updates typescript references, if required.
 - [repo] the `prepare` script now updates typescript references.
-- [core] From now on, downstream projects can refine where the configuration files (such as `settings.json`, `keymaps.json`, `recentworkspace.json`, etc.) will be stored by Theia. [#4488](https://github.com/eclipse-theia/theia/pull/4488)  
+- [core] From now on, downstream projects can refine where the configuration files (such as `settings.json`, `keymaps.json`, `recentworkspace.json`, etc.) will be stored by Theia. [#4488](https://github.com/eclipse-theia/theia/pull/4488)
```
It was on purpose. In markdown, double whitespace is a [linebreak](https://gist.github.com/shaunlebron/746476e6e7a4d698b373). Please revert. If it was automatic, we have an incorrect editorconfig somewhere.

kittaakos replied:
TIL: it seems [`\` should also work](https://gist.github.com/shaunlebron/746476e6e7a4d698b373#gistcomment-2349833). If so, please use it. Thanks!

vince-fugnitto replied:
> It was on purpose. In markdown, double whitespace is a [linebreak](https://gist.github.com/shaunlebron/746476e6e7a4d698b373). Please revert. If it was automatic, we have an incorrect editorconfig somewhere.

I reverted the change, the trimming of whitespaces was done automatically.
Unfortunately, until we find the cause it looks like this might happen each time the file is modified.

vince-fugnitto replied:
It comes from the following VS Code default preference:

<img width="411" alt="Screen Shot 2020-02-26 at 12 51 33 PM" src="https://user-images.githubusercontent.com/40359487/75372699-e3e51000-5896-11ea-86a0-3134cc08b9ec.png">

kittaakos replied:
Thanks for the update; `\` is better. 👍

vince-fugnitto replied:
> Thanks for the update; `\` is better. 👍

No problem! Thank you for catching it and providing a suggestion :)

marcdumais-work commented:
> @eclipse-theia/core
> 
> Do you think we should merge for the upcoming release (Thursday Feb 27th), or should we hold off till right after the release and potentially address any loose ends we might have missed. My understanding is that we should at least aim for before v1.0.

I would prefer so - I expect we'll be very busy at the end of next month with Theia 1.0, so if we can get this change out-of-the-way now, including the unavoidable questions about the removed extensions, it would be best. 

I'll review/test this PR, and assuming all looks good and there are no objections, we can merge early tomorrow morning, Montreal time.

P.S. we still have to review/merge one last related documentation update (https://github.com/theia-ide/theia-website/pull/89) , but that can probably be completed ASAP after the release.

marcdumais-work commented:
cleanup: The `@theia/debug` extension's [README](https://github.com/eclipse-theia/theia/blob/2cd355643242f639e0d68405edbbdac6bb18aa39/packages/debug/README.md) has a link to the removed `@theia/debug-nodejs`. The link is already broken, so I guess remove it or replace by another example? Search for "Debug adapter example for VS Code" in the References section.

marcdumais-work commented:
@vince-fugnitto I would suggest making a post on Spectrum about the removals done here. Better to do it about the same time this is merged, since extenders that build `next` version of their apps will notice soon after.

marcdumais-work commented:
cleanup: root `tsconfig.json` has a reference to removed ext `textmate-grammars`

vince-fugnitto commented:
> cleanup: The `@theia/debug` extension's [README](https://github.com/eclipse-theia/theia/blob/2cd355643242f639e0d68405edbbdac6bb18aa39/packages/debug/README.md) has a link to the removed `@theia/debug-nodejs`. The link is already broken, so I guess remove it or replace by another example? Search for "Debug adapter example for VS Code" in the References section.

I updated the docs for `@theia/debug` (I now point to the `node-debug` and `node-debug2` repos. Is this fine?

vince-fugnitto commented:
> cleanup: root `tsconfig.json` has a reference to removed ext `textmate-grammars`

I found others to clean as well, thank you for catching it, I missed removing them after rebasing often.

marcdumais-work commented:
> I now point to the `node-debug` and `node-debug2` repos. Is this fine?

It's fine to add a link for `Node Debugger 2`, but did you change the links from upstream Microsoft repos to our forks? If so please revert that part.

vince-fugnitto commented:
> > I now point to the `node-debug` and `node-debug2` repos. Is this fine?
> 
> It's fine to add a link for `Node Debugger 2`, but did you change the links from upstream Microsoft repos to our forks? If so please revert that part.

Fixed 👍

marcdumais-work commented:
As a sanity check I have built this PR, published it to a local registry and used that to build the `theia-full` application (locally outside docker) and `theia-electron`. The resulting apps looks to work fine. I think this is looking good to merge.

vince-fugnitto commented:
> LGTM, thanks Vince. Please wait tomorrow morning before merging.

Thank you for the review and help @marcdumais-work!

marcdumais-work commented:
A little thing I think we missed: the CHANGELOG part in https://github.com/eclipse-theia/theia/issues/6183#issue-493121466 . I think you should at least add in `CHANGELOG` a link pointing to your [wiki article](https://github.com/eclipse-theia/theia/wiki/Consuming-Builtin-and-External-VS-Code-Extensions). While you're in there you could add a short explanation that says the removed Theia exts are being replaced by builtins and other vs code extensions and offer the wiki link to help migrate.

We should also add an example how to use `resolutions` block to keep using Theia `0.15.0`, that still has the removed extensions (basically add to resolutions all Theia extensions pulled by the app).

vince-fugnitto commented:
> A little thing I think we missed: the CHANGELOG part in [#6183 (comment)](https://github.com/eclipse-theia/theia/issues/6183#issue-493121466) . I think you should at least add in `CHANGELOG` a link pointing to your [wiki article](https://github.com/eclipse-theia/theia/wiki/Consuming-Builtin-and-External-VS-Code-Extensions). While you're in there you could add a short explanation that says the removed Theia exts are being replaced by builtins and other vs code extensions and offer the wiki link to help migrate.
> 
> We should also add an example how to use `resolutions` block to keep using Theia `0.15.0`, that still has the removed extensions (basically add to resolutions all Theia extensions pulled by the app).

@marcdumais-work 
I included updates to the `changelog`, namely:
- adding migration steps for application developers/users to our [wiki page](https://github.com/eclipse-theia/theia/wiki/Consuming-Builtin-and-External-VS-Code-Extensions)
- an example of how to use `yarn resolutions`

Please let me know if you have any additional feedback.

marcdumais-work commented on CHANGELOG.md:
```diff
@@ -42,6 +42,34 @@ Breaking changes:
+  - The extension will no longer be maintained by the project and remains in the Git history for anyone who would like to reference it or maintain it.
+- [typescript] removed the `@theia/typescript` extension [#6933](https://github.com/eclipse-theia/theia/pull/6933)
+  - The extension will no longer be maintained by the project and remains in the Git history for anyone who would like to reference it or maintain it.
+- The release includes the removal of language-specific extensions which have been replaced by VS Code extensions.
```
Move this at the beginning of this block, so that effectively the removed extensions become the details of this item? Also we should probably mention something about the removed extensions that are not language-specific but rather are just replaced by VS Code builtins (textmate-grammars, merge-conflicts, ...). e.g.:

"The release includes the removal of language-specific Theia extensions and other Theia extensions that are or can be replaced by equivalent VS Code extensions"

marcdumais-work commented on CHANGELOG.md:
```diff
@@ -42,6 +42,34 @@ Breaking changes:
+  - The extension will no longer be maintained by the project and remains in the Git history for anyone who would like to reference it or maintain it.
+- [editorconfig] removed the `@theia/editorconfig` extension [#6933](https://github.com/eclipse-theia/theia/pull/6933)
+  - The extension will no longer be maintained by the project and remains in the Git history for anyone who would like to reference it or maintain it.
+- [java] removed the `@theia/java` extension [#6933](https://github.com/eclipse-theia/theia/pull/6933)
```
Add a mention that the theia-java app can be used as example of migrating to VS Code extensions? Add link to app under theia-apps?

marcdumais-work commented on CHANGELOG.md:
```diff
@@ -42,6 +42,34 @@ Breaking changes:
+  - The extension will no longer be maintained by the project and remains in the Git history for anyone who would like to reference it or maintain it.
+- [merge-conflicts] removed the `@theia/merge-conflicts` extension [#6933](https://github.com/eclipse-theia/theia/pull/6933)
+  - The extension will no longer be maintained by the project and remains in the Git history for anyone who would like to reference it or maintain it.
+- [python] removed the `@theia/python` extension [#6933](https://github.com/eclipse-theia/theia/pull/6933)
```
Add a mention that the theia-python app can be used as example of migrating to VS Code extensions? Add link to app under theia-apps?

marcdumais-work commented:
> I included updates to the `changelog`, namely:
> 
> * adding migration steps for application developers/users to our [wiki page](https://github.com/eclipse-theia/theia/wiki/Consuming-Builtin-and-External-VS-Code-Extensions)
> * an example of how to use `yarn resolutions`
> 
> Please let me know if you have any additional feedback.

Thanks @vince-fugnitto . I added a couple more comments in this area.

vince-fugnitto commented:
> > I included updates to the `changelog`, namely:
> > 
> > * adding migration steps for application developers/users to our [wiki page](https://github.com/eclipse-theia/theia/wiki/Consuming-Builtin-and-External-VS-Code-Extensions)
> > * an example of how to use `yarn resolutions`
> > 
> > Please let me know if you have any additional feedback.
> 
> Thanks @vince-fugnitto . I added a couple more comments in this area.

@marcdumais-work I've addressed the comments if you'd like to take a look.

vince-fugnitto commented:
@marcdumais-work feel free to merge if you are happy with the changes 😃

marcdumais-work commented on CHANGELOG.md:
````diff
@@ -42,6 +42,42 @@ Breaking changes:
+  - Alternatively, it is also possible to use [`yarn resolutions`](https://classic.yarnpkg.com/en/docs/selective-version-resolutions/) to keep using old native extensions.\
+    For example, in your application's `package.json`, you can include the following (to keep using `@theia/typescript` and `@theia/textmate-grammars`):
+    ```json
+    "resolutions": {
````
I think this might need a bit more work. e.g. If we take the browser example and try to pin a couple of compatible extensions to `0.14.0`, we'll end-up with a non-working app because we'll have a mix of `0.14.0` and `0.15.0`. 

To be sure this works, I think it's necessary to pin all theia extensions to `0.14.0`. 

Can you give this a quick try and update if needed?

akosyakov replied:
something like here with globs: https://github.com/theia-ide/theia-apps/blob/b70f8cf84830b7724acdf5f5315edef6b5b675ed/theia-ruby-docker/latest.package.json#L30 should help

akosyakov replied:
i.e. `"**/@theia/typescript": "0.15.0"`

vince-fugnitto replied:
I'm wondering, should I take out the guesswork here and just point to the `yarn resolutions` page like I've done? I can update the wiki as well to show examples with resolutions.

marcdumais-work replied:
> I'm wondering, should I take out the guesswork here and just point to the `yarn resolutions` page like I've done?

The tricky part is avoiding having multiple versions of any Theia extension being included in the app - the generic documentation will not help much with that. 

> I can update the wiki as well to show examples with resolutions.

Maybe we can test using a glob to pin all Theia extensions at once and if it works have that as example? Or alternatively just mention that it may be necessary to pin all Theia extensions and add a concrete example to the wiki ASAP.

vince-fugnitto replied:
@marcdumais-work I updated the resolution, it ended up working for me to do so.

```json
"resolutions": {
  "**/@theia/**": "0.14.0",
},
```

akosyakov replied:
It basically says use 0.14.0 for everything 😆 So maybe we just tell that since 0.15.0 these extensions are deprecated and VS Code counterparts should be used.

vince-fugnitto replied:
> It basically says use 0.14.0 for everything 😆 So maybe we just tell that since 0.15.0 these extensions are deprecated and VS Code counterparts should be used.

I'd be fine with that, what about you? @marcdumais-work

marcdumais-work replied:
0.15.0 would be better as example post-release, but would not have been a good test right now.

> It basically says use 0.14.0 for everything

True, but at least this could be used temporarily by extenders to keep building their app against 0.14.0 or 0.15.0 after we release 0.16.0, which I think would not work without the resolutions block. It can buy them time until they can adapt to use the VS Code extensions?

marcdumais-work replied:
> I'd be fine with that, what about you? @marcdumais-work

I think GH is having issues ATM - it took over an hour for my previous comment to be posted. Since we're running out of time I suggest we go with Anton's suggestion, but mentioning `0.15.0` instead of `0.14.0`, and merge ASAP. If necessary we can elaborate about keep using an older Theia version, in the wiki.

akosyakov replied:
I mean with `"**/@theia/**": "0.14.0"` there is no point to upgrade to 0.15.0, since it tells don't use 0.15.0 use 0.14.0, there is no upgrade. They could just keep using 0.14.0 as a version without any yarn resolutions.

marcdumais-work replied:
test

marcdumais-work replied:
Anton: I just tested and you're right: if I take the browser example and ask for `0.14.0` for every theia extension, the resulting app ends-up working without a resolutions block. Similar had not worked for me in the past, but if it work now, no need to document a useless work-around.

marcdumais-work commented:
Looks good, thanks @vince-fugnitto . Merging.
